engine-types/src/types/wei.rs

```diff
@@ -0,0 +1,123 @@
+use crate::fmt::Formatter;
+use crate::types::balance::error;
+use crate::types::Balance;
+use crate::{Add, Display, Sub, U256};
+
+/// Wei compatible Borsh-encoded raw value to attach an ETH balance to the transaction
+pub type WeiU256 = [u8; 32];
+
+/// Newtype to distinguish balances (denominated in Wei) from other U256 types.
+#[derive(Default, Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
+pub struct Wei(U256);
+
+impl Wei {
+    const ETH_TO_WEI: U256 = U256([1_000_000_000_000_000_000, 0, 0, 0]);
+
+    pub const fn zero() -> Self {
+        Self(U256([0, 0, 0, 0]))
+    }
+
+    pub const fn new(amount: U256) -> Self {
+        Self(amount)
+    }
+
+    // Purposely not implementing `From<u64>` because I want the call site to always
+    // say `Wei::<something>`. If `From` is implemented then the caller might write
+    // `amount.into()` without thinking too hard about the units. Explicitly writing
+    // `Wei` reminds the developer to think about whether the amount they enter is really
+    // in units of `Wei` or not.
+    pub const fn new_u64(amount: u64) -> Self {
+        Self(U256([amount, 0, 0, 0]))
+    }
+
+    pub fn from_eth(amount: U256) -> Option<Self> {
+        amount.checked_mul(Self::ETH_TO_WEI).map(Self)
+    }
+
+    pub fn to_bytes(self) -> [u8; 32] {
+        u256_to_arr(&self.0)
+    }
+
+    pub fn is_zero(&self) -> bool {
+        self.0.is_zero()
+    }
+
+    pub fn raw(self) -> U256 {
+        self.0
+    }
+
+    pub fn checked_sub(self, rhs: Self) -> Option<Self> {
+        self.0.checked_sub(rhs.0).map(Self)
+    }
+
+    pub fn checked_add(self, rhs: Self) -> Option<Self> {
+        self.0.checked_add(rhs.0).map(Self)
+    }
+
+    /// Try convert U256 to u128 with checking overflow.
+    /// NOTICE: Error can contain only overflow
+    pub fn try_into_u128(self) -> Result<u128, error::BalanceOverflowError> {
+        use crate::TryInto;
+        self.0.try_into().map_err(|_| error::BalanceOverflowError)
+    }
+}
+
+impl Display for Wei {
+    fn fmt(&self, f: &mut Formatter<'_>) -> crate::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl Add<Self> for Wei {
+    type Output = Wei;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        Self(self.0 + rhs.0)
+    }
+}
+
+impl Sub<Self> for Wei {
+    type Output = Wei;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        Self(self.0 - rhs.0)
+    }
+}
+
+/// Type casting from Wei compatible Borsh-encoded raw value into the Wei value, to attach an ETH balance to the transaction
+impl From<WeiU256> for Wei {
+    fn from(value: WeiU256) -> Self {
+        Wei(U256::from_big_endian(&value))
+    }
+}
+
+impl From<Balance> for Wei {
+    fn from(value: Balance) -> Self {
+        Wei(U256::from(value.into_u128()))
+    }
+}
```

Comment:
I'm a little concerned about existence of this implementation. I understand that inside the EVM we have the `Wei` newtype and in the connector (NEP-141 contract) we have the `Balance` newtype, and these are two representations of the same underlying asset.

However, if we are going to allow seamless converting `Balance` into `Wei` then I believe that we cannot use this `Balance` type for anything other than Aurora NEP-141 balances. For example, we should not use it to represent NEAR balances (eg when attaching NEAR to promises). It probably also needs a different name like `BridgedWei` or `NEP141Wei` or something similar so that it is clear this meant to represent that specific assent and no other 128-bit sized quantity.

Reply:
We can introduce something like:
`pub const fn new_u128(amount: u128) -> Wei {...}`

Reply:
Well, the more I think about this, the more I believe we actually should split `Balance` into multiple different newtype definitions. Right now we are using the same type for NEAR and `NEP141Wei` interchangeably which is not good. I think it would be safer if we had separate types so that we could never accidentally mix them up.

Reply:
Do you mean:
```
pub struct FungibleTokenOps<I: IO> {
    /// Total ETH supply on Near (nETH as NEP-141 token)
    pub total_eth_supply_on_near: Balance,

    /// Total ETH supply on Aurora (ETH in Aurora EVM)
    pub total_eth_supply_on_aurora: NEP141Wei,

    /// The storage size in bytes for one account.
    pub account_storage_usage: StorageUsage,
    io: I,
}
```

Reply:
And you meanl: `pub struct NEP141Wei(Wei);`?

Reply:
I think the most accurate usage of types would be something like this:

```
pub struct FungibleTokenOps<I: IO> {
    /// Total ETH supply on Near (nETH as NEP-141 token)
    pub total_eth_supply_on_near: NEP141Wei,

    /// Total ETH supply on Aurora (ETH in Aurora EVM)
    pub total_eth_supply_on_aurora: Wei,

    /// The storage size in bytes for one account.
    pub account_storage_usage: StorageUsage,
    io: I,
}
```

Because it uses the NEP-141 type for the quantity on NEAR and the EVM type for the quantity in the engine.

And definition of `NEP141Wei` would be `pub struct NEP141Wei(u128)`.

Additionally there would be another type, say `pub struct Yocto(u128)` (name inspired by the fact 1 NEAR = 10^24 yoctoNEAR, and 1 yoctoNEAR is the smallest unit), and this would be used in places where we expect NEAR tokens to be. For example: attached deposit for promises and storage registration.

Reply:
One notice here. Currently, for `ft_transfer_call` we have input data:
```
pub struct TransferCallCallArgs {
    pub receiver_id: AccountId,
    pub amount: Balance,
    pub memo: Option<String>,
    pub msg: String,
}
```
Where `amount: Balance`. It's same amount for `eth_on_aurora` & `eth_on_near`.
I think we can't change it to `U256` due to incompatibility with the current format.

For other cases we can improve `Balances` types with your proposal.

Reply:
Need to `impl From<Balance> for Wei` now that `Balance` definitely does not represent any sort of Wei.